File: devices-api-java-mvc/src/main/java/com/lucas/devicesapijavamvc/kafka/DeviceEventPublisher.java
package com.lucas.devicesapijavamvc.kafka;

import com.lucas.devicesapijavamvc.domain.Device;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

@Component
@ConditionalOnProperty(value = "kafka.enabled", havingValue = "true")
public class DeviceEventPublisher {

    private final KafkaTemplate<String, DeviceEvent> template;
    private final String topic;

    public DeviceEventPublisher(
            KafkaTemplate<String, DeviceEvent> template,
            @Value("${kafka.topic.devices}") String topic
    ) {
        this.template = template;
        this.topic = topic;
    }

    public void publishCreated(Device device, String origin) {
        DeviceEvent event = new DeviceEvent(
                device.getExternalId(),
                device.getName(),
                device.getBrand(),
                device.getState().name(),
                device.getCreationTime(),
                origin
        );
        template.send(topic, device.getExternalId().toString(), event);
    }
}

File: devices-api-kotlin-reactive/src/main/kotlin/com/lucas/devicesapikotlinreactive/infrastructure/messaging/DeviceEventPublisher.kt
package com.lucas.devicesapikotlinreactive.infrastructure.messaging

import com.lucas.devicesapikotlinreactive.domain.model.Device
import com.lucas.devicesapikotlinreactive.domain.port.DeviceEventPublisherPort
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
@ConditionalOnProperty(prefix = "kafka", name = ["enabled"], havingValue = "true")
class DeviceEventPublisher(
    private val kafkaTemplate: KafkaTemplate<String, DeviceEvent>,
    private val kafkaProps: KafkaProperties
) : DeviceEventPublisherPort {

    override fun publishCreated(device: Device): Mono<Void> {
        val event = device.toEvent(origin = "KOTLIN")

        return Mono.fromRunnable {
            kafkaTemplate.send(
                kafkaProps.topic.devices,
                event.externalId.toString(),
                event
            )
        }
    }

}
